java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java

```diff
@@ -2292,10 +2292,11 @@ private void readDictionaryStream(InStream in) throws IOException {
           int dictionaryBufferSize = dictionaryOffsets[dictionaryOffsets.length - 1];
           dictionaryBuffer = new byte[dictionaryBufferSize];
           int pos = 0;
-          int chunkSize = in.available();
-          byte[] chunkBytes = new byte[chunkSize];
+          //check if dictionary size is smaller than available stream size to avoid ArrayIndexOutOfBoundsException
+          int readSize = Math.min(in.available(), dictionaryBufferSize);
+          byte[] chunkBytes = new byte[readSize];
```

Comment:
The previous reading code was implemented on the premise that the total amount of data in the stream is the same as the expected amount to be read. I am curious as to why `in.available()` is greater than `dictionaryBufferSize` in your environment.

Reply:
I'm also curious, `available `currently returns the remaining value of the Bytebuffer, could you provide the specific scenario?
@zratkai

Reply:
In Hive the ORC's Reader classes (e.g. StringTreeReader)  have sublclasses and ORC file is read with LLAP. LLAP uses a default 4kB blocks to read, and this causes that the in.available() can be bigger than dictionary size. (If dictionary is smaller than 4kB.) This fix will not cause any issue if ORC file is read without LLAP, only with ORC classes, since in System.arraycopy() the maximum amount which can be copied can not be bigger than destination or source size. The previous reading mode created a bigger destination array in Hive than necessary.

Reply:
@zratkai Thank you for your detailed explanation. I understand your use case.

I think `in.available()` will be recalculated every time a new block is read, right? So this fix seems not perfect.
For example, at the beginning,
`in.available()` = 4k
`dictionaryBufferSize` = 5k
`readSize` = 4k
Entering the loop,
the first time` currentLength` is read as 4k, and then copied.
The second time `currentLength` may still be 4k, which will still cause an out of bounds error.